src/sampleutil.cpp

```diff
@@ -17,19 +17,78 @@ typedef qint32 int32_t;
 // When changing, be carefull to not prevent the vectorizing
 // https://gcc.gnu.org/projects/tree-ssa/vectorization.html
 
+// TODO() Check if uintptr_t is availabe on all our build targets and use that
+// instead of size_t, we can remove the sizeof(size_t) check than 
+static inline bool useAlignedAlloc() {
+    // This will work on all targets and compilers. 
+    // It will return true on MSVC 32 bit builds and false for
+    // Linux 32 and 64 bit builds
+    return (sizeof(long double) == 8 && sizeof(CSAMPLE*) <= 8 &&
+            sizeof(CSAMPLE*) == sizeof(size_t));
+}
+
 // static
-CSAMPLE* SampleUtil::alloc(unsigned int size) {
-    // TODO(XXX) align the array
-    return new CSAMPLE[size];
+CSAMPLE* SampleUtil::alloc(int size) {
+    // For optimal use of SSE registers, it is required to align
+    // the sample buffers to 16 Byte (128 bit) boundary
+
+    // Pointers returned by malloc are aligned for the largest scalar type,
+    // which is long double with usually 16 byte.
+    // An exception is MSVC X86 where long double is mapped to double.
+
+    // In case of sizeof(long double) = 8
+    // this code over allocates the requested buffer to be able to
+    // shift the returned pointer to a 16 byte alignment
+    // In the memory before, a pointer to the original
+    // malloced area is stored used to free the memory
+    // This code can be replaced by C11 <stdlib.h> aligned_alloc()
+    // or MSVC ::_aligned_malloc(size, alignment) and  ::_aligned_free(ptr);
+
+    if (useAlignedAlloc()) {
+#if(_MSC_VER)
+        return (CSAMPLE*)_aligned_malloc(size, 16);
+#else
+        // This block will be only used on exotic builds         
+        // We need to shift the alignment to 16
+        const size_t alignment = 16;
+        const size_t unaligned_size = sizeof(CSAMPLE[size]) + alignment;
+        void* pUnaligned = std::malloc(unaligned_size);
+        if (pUnaligned == NULL) {
+            return NULL;
+        }
+        // Shift
+        void* pAlligned = (void*)(((size_t)pUnaligned & ~(alignment - 1)) + alignment);
+        // Store pointer to original relative to the shifted pointer
+        *((void**)(pAlligned) - 1) = pUnaligned;
+        return (CSAMPLE*)pAlligned;
+#endif
+    } else {
+        // We are either correct aligned or on an exotic architecture
+        return new CSAMPLE[size];
+    }
 }
 
 void SampleUtil::free(CSAMPLE* pBuffer) {
-    delete[] pBuffer;
+    // See SampleUtil::alloc() for details
+    if (useAlignedAlloc()) {
+        if (pBuffer == NULL) {
+            return;
+        }
+#if(_MSC_VER)
+        _aligned_free(pBuffer);
+#else
+        // Pointer to the original memory is stored before pBuffer
+        std::free(*((void**)((void*)pBuffer) - 1));
```

Comment:
This line causes an error with clang on OSX. I cannot reproduce it now since my MBP is temporarily out of commission, but it has to do with pBuffer being a CSAMPLE *.